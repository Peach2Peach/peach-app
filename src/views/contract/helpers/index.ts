export { isPSBTForBatch } from "./isPSBTForBatch";
export { shouldShowTradeStatusInfo } from "./shouldShowTradeStatusInfo";
export { verifyReleasePSBT as verifyPSBT } from "./verifyReleasePSBT";
