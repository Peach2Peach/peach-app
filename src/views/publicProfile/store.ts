import create from 'zustand'

type ProfileStore = {
  currentUserPubkey: string
  setCurrentUserPubkey: (currentUserPubkey: string) => void
}

export const useProfileStore = create<ProfileStore>()((set) => ({
  currentUserPubkey: '',
  setCurrentUserPubkey: (currentUserPubkey: string) => set({ currentUserPubkey }),
}))
