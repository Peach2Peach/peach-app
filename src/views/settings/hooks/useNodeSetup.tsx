import { shallow } from 'zustand/shallow'
import { useToggleBoolean, useValidatedState } from '../../../hooks'
import { useWalletState } from '../../../utils/wallet/walletStore'

const addressRules = { required: true, url: true }
export const useNodeSetup = () => {
  const [node, setCustomNode] = useWalletState((state) => [state.node, state.setCustomNode], shallow)
  const [enabled, toggleEnabled] = useToggleBoolean(node.enabled)
  const [ssl, toggleSSL] = useToggleBoolean(node.ssl)
  const [address, setAddress, isAddressValid, addressErrors] = useValidatedState<string>(
    node.address || '',
    addressRules,
  )
  const canCheckConnection = enabled && isAddressValid
  const isConnected = false

  const checkConnection = () => {}
  const save = () => setCustomNode({ enabled, ssl, address })

  return {
    enabled,
    toggleEnabled,
    ssl,
    toggleSSL,
    isConnected,
    address,
    setAddress,
    addressErrors,
    canCheckConnection,
    checkConnection,
  }
}
