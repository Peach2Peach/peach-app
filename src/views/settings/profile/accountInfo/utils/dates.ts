import i18n from '../../../../../utils/i18n'

export const getTimeDiffInDays = (date: Date) => Math.floor((Date.now() - new Date(date).getTime()) / (86400 * 1000))

export const getDateToDisplay = (date: Date) => {
  const newDate = new Date(date)
  const dateString = newDate.toLocaleDateString('en-GB')
  const numberOfDays = getTimeDiffInDays(newDate)
  const daysAgo
    = numberOfDays > 1
      ? i18n('profile.daysAgo', String(getTimeDiffInDays(newDate)))
      : numberOfDays === 1
        ? i18n('profile.yesterday')
        : i18n('profile.today')

  return `${dateString} (${daysAgo})`
}
