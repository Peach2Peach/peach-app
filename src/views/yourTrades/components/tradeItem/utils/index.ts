export { getActionIcon } from './getActionIcon'
export { getActionLabel } from './getActionLabel'
