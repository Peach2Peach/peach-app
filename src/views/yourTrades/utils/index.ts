export { getDisputeResultTheme } from './getDisputeResultTheme'
export { getNavigationDestinationForContract, getNavigationDestinationForOffer } from './getNavigationDestination'
export { getOfferLevel } from './getOfferLevel'
export { getThemeForPastTrade } from './getThemeForPastTrade'
export { hasDoubleMatched } from './hasDoubleMatched'
export { isContractSummary } from './isContractSummary'
export { isOpenAction } from './isOpenAction'
export { isOpenOffer } from './isOpenOffer'
export { isPastOffer } from './isPastOffer'
export { isPrioritary } from './isPrioritary'
export { isWaiting } from './isWaiting'
export { navigateToContract } from './navigateToContract'
export { navigateToOffer } from './navigateToOffer'
export { shouldGoToOfferSummary } from './shouldGoToOfferSummary'
export { statusIcons } from './statusIcons'
export { wonDisputeForTrade } from './wonDisputeForTrade'
