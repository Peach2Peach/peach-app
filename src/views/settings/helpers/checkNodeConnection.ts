import { Blockchain } from 'bdk-rn'
import { BlockChainNames, BlockchainElectrumConfig, BlockchainEsploraConfig } from 'bdk-rn/lib/lib/enums'
import { info } from '../../../utils/log'
import { parseError } from '../../../utils/result'
import { addProtocol } from '../../../utils/web'

const checkElectrumConnection = async (address: string, ssl: boolean) => {
  const config: BlockchainElectrumConfig = {
    url: addProtocol(address, ssl ? 'ssl' : 'tcp'),
    sock5: null,
    retry: 1,
    timeout: 5,
    stopGap: 1,
    validateDomain: false,
  }

  try {
    info('Checking electrum connection...')
    const blockchain = await new Blockchain().create(config, BlockChainNames.Electrum)
    await blockchain.getBlockHash()
    return { result: BlockChainNames.Electrum }
  } catch (e) {
    info('electrum connection failed')
    return { error: parseError(e) }
  }
}
const checkEsploraConnection = async (address: string, ssl: boolean) => {
  const config: BlockchainEsploraConfig = {
    baseUrl: addProtocol(address, ssl ? 'https' : 'http'),
    proxy: null,
    concurrency: 1,
    timeout: 5,
    stopGap: 1,
  }

  try {
    info('Checking esplora connection...')
    const blockchain = await new Blockchain().create(config, BlockChainNames.Esplora)
    await blockchain.getBlockHash()
    return { result: BlockChainNames.Esplora }
  } catch (e) {
    info('esplora connection failed')
    return { error: parseError(e) }
  }
}
const connectionChecks = [checkElectrumConnection, checkEsploraConnection]
export const checkNodeConnection = async (address: string, ssl = false) => {
  const errors: unknown[] = []
  for (const connectionCheck of connectionChecks) {
    // eslint-disable-next-line no-await-in-loop
    const { result, error } = await connectionCheck(address, ssl)
    if (result) return { result }
    errors.push(error)
  }
  return { error: errors.join('\n\n') }
}
