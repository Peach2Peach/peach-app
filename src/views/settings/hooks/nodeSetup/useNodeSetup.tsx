import { BlockChainNames } from 'bdk-rn/lib/lib/enums'
import { useEffect, useState } from 'react'
import { shallow } from 'zustand/shallow'
import { PeachText } from '../../../../components/text/PeachText'
import { useShowLoadingPopup } from '../../../../hooks/useShowLoadingPopup'
import { useToggleBoolean } from '../../../../hooks/useToggleBoolean'
import { useValidatedState } from '../../../../hooks/useValidatedState'
import { SuccessPopup } from '../../../../popups/SuccessPopup'
import { WarningPopup } from '../../../../popups/WarningPopup'
import { ClosePopupAction } from '../../../../popups/actions/ClosePopupAction'
import { LoadingPopupAction } from '../../../../popups/actions/LoadingPopupAction'
import { usePopupStore } from '../../../../store/usePopupStore'
import tw from '../../../../styles/tailwind'
import i18n from '../../../../utils/i18n'
import { useNodeConfigState } from '../../../../utils/wallet/nodeConfigStore'
import { peachWallet } from '../../../../utils/wallet/setWallet'
import { checkNodeConnection } from '../../helpers/checkNodeConnection'

const urlRules = { required: true, url: true }
export const useNodeSetup = () => {
  const showLoadingPopup = useShowLoadingPopup()
  const setPopup = usePopupStore((state) => state.setPopup)

  const [node, setCustomNode, enabled, toggleEnabled] = useNodeConfigState(
    (state) => [state, state.setCustomNode, state.enabled, state.toggleEnabled],
    shallow,
  )
  const [ssl, toggleSSL] = useToggleBoolean(node.ssl)
  const [url, setURL, isURLValid, urlErrors] = useValidatedState<string>(node.url || '', urlRules)
  const canCheckConnection = enabled && isURLValid
  const [isConnected, setIsConnected] = useState(!!node.url)

  const editConfig = () => setIsConnected(false)
  const save = (blockchainType: BlockChainNames) => {
    setCustomNode({ enabled, ssl, url, type: blockchainType })
    setIsConnected(true)
    peachWallet.setBlockchain({ enabled, ssl, url, type: blockchainType })
    peachWallet.initWallet()
  }

  const checkConnection = async () => {
    showLoadingPopup({ title: i18n('wallet.settings.node.checkingConnection') })

    const { result: nodeType, error } = await checkNodeConnection(url, ssl)
    if (nodeType) {
      return setPopup(<NodeConnectionSuccessPopup url={url} save={() => save(nodeType)} />)
    }
    return setPopup(<NodeConnectionErrorPopup error={error} />)
  }

  useEffect(() => {
    peachWallet.setBlockchain(node)
  }, [node])

  return {
    enabled,
    toggleEnabled,
    ssl,
    toggleSSL,
    isConnected,
    url,
    setURL,
    urlErrors,
    editConfig,
    canCheckConnection,
    checkConnection,
  }
}

type ErrorPopupProps = {
  error: string
}

function NodeConnectionErrorPopup ({ error }: ErrorPopupProps) {
  return (
    <WarningPopup
      title={i18n('wallet.settings.node.error.title')}
      content={<PeachText selectable>{i18n('wallet.settings.node.error.text', error)}</PeachText>}
      actions={<ClosePopupAction style={tw`justify-center`} textStyle={tw`text-black-100`} />}
    />
  )
}

type SuccessPopupProps = {
  url: string
  save: () => void
}

function NodeConnectionSuccessPopup ({ url, save }: SuccessPopupProps) {
  return (
    <SuccessPopup
      title={i18n('wallet.settings.node.success.title')}
      content={<PeachText>{i18n('wallet.settings.node.success.text', url)}</PeachText>}
      actions={
        <>
          <ClosePopupAction />
          <SaveAction {...{ save }} />
        </>
      }
    />
  )
}

function SaveAction ({ save }: Pick<SuccessPopupProps, 'save'>) {
  const closePopup = usePopupStore((state) => state.closePopup)
  const onPress = () => {
    save()
    closePopup()
  }

  return (
    <LoadingPopupAction
      onPress={onPress}
      label={i18n('wallet.settings.node.success.confirm')}
      iconId={'save'}
      reverseOrder
    />
  )
}
