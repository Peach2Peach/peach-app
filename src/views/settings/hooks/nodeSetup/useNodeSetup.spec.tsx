import { act, fireEvent, render, renderHook } from '@testing-library/react-native'
import { BlockChainNames } from 'bdk-rn/lib/lib/enums'
import { NavigationWrapper } from '../../../../../tests/unit/helpers/NavigationWrapper'
import { PopupLoadingSpinner } from '../../../../../tests/unit/helpers/PopupLoadingSpinner'
import { usePopupStore } from '../../../../store/usePopupStore'
import { getError, getResult } from '../../../../utils/result'
import { PeachWallet } from '../../../../utils/wallet/PeachWallet'
import { NodeConfig, useNodeConfigState } from '../../../../utils/wallet/nodeConfigStore'
import { setPeachWallet } from '../../../../utils/wallet/setWallet'
import { useNodeSetup } from './useNodeSetup'

const checkNodeConnectionMock = jest.fn()
jest.mock('../../helpers/checkNodeConnection', () => ({
  checkNodeConnection: (...args: unknown[]) => checkNodeConnectionMock(...args),
}))

describe('useNodeSetup', () => {
  const url = 'blockstream.info'
  beforeEach(() => {
    // @ts-expect-error mock doesn't need args
    setPeachWallet(new PeachWallet())
    useNodeConfigState.getState().reset()
  })
  it('should return defaults', () => {
    const { result } = renderHook(useNodeSetup)
    expect(result.current).toEqual({
      enabled: false,
      toggleEnabled: expect.any(Function),
      ssl: false,
      toggleSSL: expect.any(Function),
      isConnected: false,
      url: '',
      urlErrors: ['this field is required', 'please enter a valid value'],
      setURL: expect.any(Function),
      canCheckConnection: false,
      checkConnection: expect.any(Function),
      editConfig: expect.any(Function),
    })
  })
  it('should load setup from nodeConfigState', () => {
    const nodeSetup: NodeConfig = {
      enabled: true,
      ssl: true,
      url,
      type: BlockChainNames.Electrum,
    }
    useNodeConfigState.getState().setCustomNode(nodeSetup)
    const { result } = renderHook(useNodeSetup)
    expect(result.current).toEqual({
      ...result.current,
      ...nodeSetup,
      type: undefined,
      canCheckConnection: true,
    })
  })
  it('should set url', () => {
    const { result } = renderHook(useNodeSetup)
    act(() => result.current.setURL(url))
    expect(result.current.url).toBe(url)
  })
  it('should validate url', () => {
    const { result } = renderHook(useNodeSetup)
    act(() => {
      result.current.toggleEnabled()
      result.current.setURL('https://')
    })
    expect(result.current.urlErrors).toEqual(['please enter a valid value'])
    expect(result.current.canCheckConnection).toBeFalsy()
  })
  it('canCheckConnection is true when enabled and url is configured and valid', () => {
    const { result } = renderHook(useNodeSetup)
    act(() => {
      result.current.toggleEnabled()
      result.current.setURL(url)
    })
    expect(result.current.canCheckConnection).toBeTruthy()
  })
  it('should successfully check connection and save config', async () => {
    const promise = Promise.resolve(getResult('esplora'))
    checkNodeConnectionMock.mockReturnValue(promise)
    const { result } = renderHook(useNodeSetup)
    act(() => {
      result.current.toggleEnabled()
      result.current.setURL(url)
    })
    act(() => {
      result.current.checkConnection()
    })

    expect(checkNodeConnectionMock).toHaveBeenCalledWith(url, false)
    expect(usePopupStore.getState()).toEqual({
      ...usePopupStore.getState(),
      action1: {
        callback: expect.any(Function),
        icon: 'clock',
        label: 'loading...',
      },
      closePopup: usePopupStore.getState().closePopup,
      content: PopupLoadingSpinner,
      level: 'APP',
      title: 'checking connection',
      visible: true,
    })
    await act(async () => {
      await promise
    })
    const popup = usePopupStore.getState().popupComponent || <></>
    const { getByText } = render(popup, { wrapper: NavigationWrapper })
    const button = getByText('save node info')
    fireEvent.press(button)
    expect(result.current.isConnected).toBeTruthy()
    expect(useNodeConfigState.getState()).toEqual({
      ...useNodeConfigState.getState(),
      enabled: true,
      url,
      type: 'esplora',
    })
  })
  it('should show connection error popup if no connection could be made', async () => {
    const error = 'error'
    const promise = Promise.resolve(getError(error))
    checkNodeConnectionMock.mockReturnValue(promise)
    const { result } = renderHook(useNodeSetup)
    act(() => {
      result.current.toggleEnabled()
      result.current.setURL(url)
    })
    act(() => {
      result.current.checkConnection()
    })

    expect(checkNodeConnectionMock).toHaveBeenCalledWith(url, false)

    await act(async () => {
      await promise
    })
    const popup = usePopupStore.getState().popupComponent || <></>
    expect(usePopupStore.getState().visible).toBe(true)
    const { toJSON } = render(popup, { wrapper: NavigationWrapper })
    expect(toJSON()).toMatchSnapshot()
  })
  it('should should set connected to false when calling editConfig', () => {
    const nodeSetup: NodeConfig = {
      enabled: true,
      ssl: true,
      url,
      type: BlockChainNames.Electrum,
    }
    useNodeConfigState.getState().setCustomNode(nodeSetup)
    const { result } = renderHook(useNodeSetup)
    expect(result.current.isConnected).toBeTruthy()
    act(() => result.current.editConfig())
    expect(result.current.isConnected).toBeFalsy()
  })
})
