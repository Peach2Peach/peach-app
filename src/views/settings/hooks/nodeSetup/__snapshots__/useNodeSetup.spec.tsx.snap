// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useNodeSetup should show connection error popup if no connection could be made 1`] = `
<View
  style={
    {
      "borderRadius": 16,
      "marginLeft": 24,
      "marginRight": 24,
      "overflow": "hidden",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "stretch",
          "backgroundColor": "#FEEDE5",
          "gap": 12,
          "paddingBottom": 24,
          "paddingLeft": 24,
          "paddingRight": 24,
          "paddingTop": 16,
        },
        [
          {
            "backgroundColor": "#FDF6C0",
          },
          {
            "alignItems": "stretch",
          },
        ],
      ]
    }
  >
    <Text
      allowFontScaling={false}
      style={
        [
          {
            "color": "#2B1911",
            "fontFamily": "Baloo2-Regular",
            "fontSize": 16,
            "letterSpacing": 0.16,
            "lineHeight": 24,
          },
          {
            "fontFamily": "Baloo2-Bold",
            "fontSize": 24,
            "letterSpacing": 0.24,
            "lineHeight": 32,
            "width": "100%",
          },
          false,
        ]
      }
    >
      no connection...
    </Text>
    <Text
      allowFontScaling={false}
      selectable={true}
      style={
        [
          {
            "color": "#2B1911",
            "fontFamily": "Baloo2-Regular",
            "fontSize": 16,
            "letterSpacing": 0.16,
            "lineHeight": 24,
          },
          undefined,
          false,
        ]
      }
    >
      Please check if you have entered all the connection information correctly.

We got the followingÂ response:

error
    </Text>
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "alignSelf": "stretch",
          "backgroundColor": "#F56522",
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
        {
          "backgroundColor": "#F5CE22",
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "width": "100%",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "flexGrow": 1,
            "gap": 4,
            "opacity": 1,
            "paddingBottom": 8,
            "paddingLeft": 24,
            "paddingRight": 24,
            "paddingTop": 8,
          }
        }
      >
        <test-file-stub
          fill="#2B1911"
          style={
            [
              {
                "height": 16,
                "width": 16,
              },
              undefined,
            ]
          }
        />
        <Text
          allowFontScaling={false}
          style={
            [
              {
                "color": "#2B1911",
                "fontFamily": "Baloo2-Regular",
                "fontSize": 16,
                "letterSpacing": 0.16,
                "lineHeight": 24,
              },
              [
                {
                  "color": "#FFFCFA",
                  "fontFamily": "Baloo2-SemiBold",
                  "fontSize": 16,
                  "letterSpacing": 0.16,
                  "lineHeight": 28,
                },
                {
                  "color": "#2B1911",
                },
              ],
              false,
            ]
          }
        >
          close
        </Text>
      </View>
    </View>
  </View>
</View>
`;
